### PR TITLE
handle a list as source in Copy

### LINKED ISSUE #3009: Copy(somedir, Glob('otherdir/*')) fails because src list is dropped
This issue was originally created at: 2015-07-27 06:17:29.
This issue was reported by: `knut3`.
knut3 said at 2015-07-27 06:17:29
>Hi,
> 
I want to copy all files from a directory to another in a builder. I use:
 
```py
Execute(Copy('somedir', Glob('otherdir/*')))
```
 
This fails with error: `No such file or directory`
 
`Copy()` is implemented like this:
 
```py
def copy_func(dest, src):
    SCons.Node.FS.invalidate_node_memos(dest)
    if SCons.Util.is_List(src) and os.path.isdir(dest):
        for file in src:
            shutil.copy2(file, dest)
        return 0
    elif os.path.isfile(src):
        return shutil.copy2(src, dest)
    else:
        return shutil.copytree(src, dest, 1)
 
Copy = ActionFactory(copy_func,
                     lambda dest, src: 'Copy("%s", "%s")' % (dest, src),
                     convert=str)
```
 
---
 
I added some prints to see what's happening here:
 
```py
SCons.Util.is_List(src) => false
type(src) => str
```
 
`src` is converted to str by mistake. Thus, the wrong if clause is taken in `copy_func()`. This is caused by `convert=str` of the `ActionFactory`. If I remove this, scons fails somewhere else, but the list is preserved correctly.

### PR DESCRIPTION
**Pull request by mwichmann**
`Copy` (in the `copy_func()`) unconditionally converts the source argument to a string.  This breaks the case where it was passed as a list - it should remain a list with the members converted to string, rather than being a single string. We assume a list source is supported, since the code just a few lines lower down detects a list and iterates through it. A failure is added if a copy of a list is detected to an existing filesystem entry that is not a directory. A test is added in `test/Copy-Action.py` to try to detect the copy-list case.

With the change to make a list argument work, found that the message printed by the matching strfunction (originally a lambda) needed the same adjustment, if the caller did not supply a list that was already strings (e.g. a list of Nodes, as from calling `Glob()`). Fixed up.

At the same time, did some fiddling with the other Action Functions.  `Chmod` had not been adjusted to emit the new Python standard for octal values (`0o755` instead of `0755`) - the change is only for the message, not for the function, but required aligning the test with the new format.

Added some docstrings.

Fixes #3009

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `CHANGES.txt` (and read the `README.rst`)
* [X] I have updated the appropriate documentation

### CODE REVIEW

File: SCons/Defaults.py [hunk 1]
```diff
@@ -231,55 +242,84 @@ def chmod_func(dest, mode):
...
         for file in src:
             shutil.copy2(file, dest)
         return 0
+
     elif os.path.islink(src):
         if symlinks:
-            return os.symlink(os.readlink(src), dest)
-        else:
-            return copy_func(dest, os.path.realpath(src))
+            os.symlink(os.readlink(src), dest)
```
Comment (bdbaddog) [inline SCons/Defaults.py:296]: why not return os.symlinks()'s return value?

File: CHANGES.txt [hunk 2]
```diff
@@ -27,6 +27,13 @@ RELEASE  VERSION/DATE TO BE FILLED IN LATER
   From Ryan Saunders:
     - Fixed runtest.py failure on Windows caused by excessive escaping of the path to python.exe.
 
+  From Mats Wichmann:
+    - A list argument as the source to the Copy() action function is now
+      correctly handled by converting elements to string. Copy bails if
```
Comment (bdbaddog) [inline CHANGES.txt]: maybe instead of `bails` will throw a `SCons.Errors.UserError` ?

File: SCons/Defaults.py (hunk 1 above)
Comment (mwichmann) [reply SCons/Defaults.py:296]: was trying to be a bit more consistent -` os.symlink` doesn't actually return (aka None) and this *seems* to want to return 0 on success.  except I'm far from convinced anything looks at the return from one of these functions the way they're embedded in whatever an ActionFactor sets up.

File: CHANGES.txt (hunk 2 above)
Comment (mwichmann) [reply CHANGES.txt]: yes, ok - I'm going to change this, so it won't be a `UserError` anyway, but will adjust the wording.

File: SCons/Defaults.py (hunk 1 above)
Comment (bdbaddog) [reply SCons/Defaults.py:296]: Gotcha.

File: CHANGES.txt (hunk 2 above)
Comment (bdbaddog) [reply CHANGES.txt]: Gotcha. Too bad there's nothing which gets called at SConscript evaluation time for Actions which could validate this and fail then..

**Comment by mwichmann** [comment]
I rebased this to back out changes to four doc (xml) files that actually belong to a different PR.

**Pull request merged**